File: 0395-longest-substring-with-at-least-k-repeating-characters/0395-longest-substring-with-at-least-k-repeating-characters.java
class Solution {
    public int longestSubstring(String s, int k) {
        char[] arr = s.toCharArray();
        int n = s.length();
        
        HashMap<Character,Integer> mp = new HashMap<>();
        
        for(char it:arr){
            if(mp.containsKey(it)){
                mp.put(it, mp.get(it) + 1);
            }
            else{
                mp.put(it, 1);
            }
        }
        
        int idx=0;
        while(idx < n && mp.get(arr[idx]) >= k)idx++;
        
        if(idx==n)return n;
        
        int left = longestSubstring(s.substring(0,idx),k);
        int right = longestSubstring(s.substring(idx+1),k);
        
        return Math.max(left,right);
    }
}

File: 0395-longest-substring-with-at-least-k-repeating-characters/0395-longest-substring-with-at-least-k-repeating-characters.cpp
class Solution {
public:
    int longestSubstring(string s, int k) {
        if(k>s.size() || s.size()==0) return 0;
        if(k==0)return s.size();

        unordered_map<char,int>mp;
        
        for(auto it:s){
            mp[it]++;
        }
        
        int idx=0;
        while(idx<s.size() && mp[s[idx]] >= k)idx++;
        
        if(idx==s.size())return s.size();
        
        int left = longestSubstring(s.substr(0,idx),k);
        int right = longestSubstring(s.substr(idx+1),k);
        
        return max(left,right);
    }
};